java/kotlin-extractor2/src/main/kotlin/entities/Expression.kt

```diff
@@ -215,25 +221,57 @@ OLD: KE1
         }
 */
 
-private fun isFunction(
+private fun isFunctionMatchingNames(
+    symbol: KaFunctionSymbol,
+    name: CallableId,
+    isExtension: Boolean = false,
+    nullability: KaTypeNullability = KaTypeNullability.UNKNOWN,
+    extensionReceiverClassName: ClassId? = null
+): Boolean {
+
+    if (symbol.callableId != name)
+        return false
+
+    if (!isExtension)
+        return true
+
+    val receiverType = symbol.receiverParameter?.type
+    if (receiverType == null)
+        return false
+
+    if (receiverType.nullability != nullability)
+        return false
+
+    if (receiverType.symbol?.classId != extensionReceiverClassName)
+        return false
+
+    return true
+}
+
+private fun isFunctionMatchingName(
     symbol: KaFunctionSymbol,
     packageName: String,
-    className: String,
+    className: String?,
     functionName: String
 ): Boolean {
 
-    return symbol.callableId?.packageName?.asString() == packageName &&
-            symbol.callableId?.className?.asString() == className &&
-            symbol.callableId?.callableName?.asString() == functionName
+    return isFunctionMatchingNames(
+        symbol,
+        CallableId(
```

Comment:
I think that it would be better to avoid constructing anything from kotlinc if possible, especially if we are aiming to use concurrency, as then we don't have to worry about what side-effects etc there might be. Here I think we can just as easily destruct the CallableId at the other end, as construct it here.

Reply:
`CallableId` and `ClassId` are tuples of 3 and 2 names. Should we introduce custom data classes instead of reusing the kotlin compiler types? I don't think we'd want to pass 3 + 2 String args instead.

Reply:
I've just been looking at how we might handle this problem in general; I'm more interested in cases where we want to extract e.g. a generated function, for which the analysis API doesn't give us the AST and we have to fabricate it. I was hoping that we could have our own CodeQlFunction interface, and have it be implemented by KtFunction using something like delegates. However, it looks like that isn't possible; there have long been discussions and proposals (e.g. https://youtrack.jetbrains.com/issue/KT-505, https://github.com/Kotlin/KEEP/pull/87) but no implementation yet.

Perhaps `CallableId` and `ClassId` are simple enough that we should just reuse them? Or at least, do that for now until we come to something like the generated-function problem?